python/ray/serve/tests/test_cancellation.py

```diff
@@ -0,0 +1,418 @@
+import asyncio
+import concurrent.futures
+import pytest
+import sys
+
+from fastapi import FastAPI
+import requests
+from starlette.requests import Request
+
+import ray
+from ray.actor import ActorHandle
+from ray._private.test_utils import (
+    SignalActor,
+    async_wait_for_condition,
+    wait_for_condition,
+)
+
+from ray import serve
+from ray.serve._private.constants import RAY_SERVE_ENABLE_NEW_ROUTING
+
+
+async def send_signal_on_cancellation(signal_actor: ActorHandle):
+    try:
+        await asyncio.sleep(100000)
+    except asyncio.CancelledError:
+        await signal_actor.send.remote()
+
+
+@pytest.mark.skipif(
+    not RAY_SERVE_ENABLE_NEW_ROUTING,
+    reason="New routing feature flag is disabled.",
+)
+@pytest.mark.parametrize("use_fastapi", [False, True])
+def test_cancel_on_http_client_disconnect_during_execution(
+    serve_instance, use_fastapi: bool
+):
+    """Test the client disconnecting while the handler is executing."""
+    inner_signal_actor = SignalActor.remote()
+    outer_signal_actor = SignalActor.remote()
+
+    @serve.deployment
+    async def inner():
+        await send_signal_on_cancellation(inner_signal_actor)
+
+    if use_fastapi:
+        app = FastAPI()
+
+        @serve.deployment
+        @serve.ingress(app)
+        class Ingress:
+            def __init__(self, handle):
+                self._handle = handle.options(use_new_handle_api=True)
+
+            @app.get("/")
+            async def wait_for_cancellation(self):
+                await self._handle.remote()._to_object_ref()
+                await send_signal_on_cancellation(outer_signal_actor)
+
+    else:
+
+        @serve.deployment
+        class Ingress:
+            def __init__(self, handle):
+                self._handle = handle.options(use_new_handle_api=True)
+
+            async def __call__(self, request: Request):
+                await self._handle.remote()._to_object_ref()
+                await send_signal_on_cancellation(outer_signal_actor)
+
+    serve.run(Ingress.bind(inner.bind()))
+
+    # Intentionally time out on the client, causing it to disconnect.
+    with pytest.raises(requests.exceptions.ReadTimeout):
+        requests.get("http://localhost:8000", timeout=0.5)
+
+    # Both the HTTP handler and the inner deployment handle call should be cancelled.
+    ray.get(inner_signal_actor.wait.remote())
```

Comment:
Could we switch these to `ray.wait(..., timeout=0.5)` calls? Otherwise, if these calls hang, the test will fail only once the entire test suite times out, and none of the other tests will run.

This pattern is used throughout a few of these tests. Could we switch all of them to use `ray.wait` as well?

Reply:
yea done